File: le1.go
package leetcode

import "sort"

// https://leetcode-cn.com/problems/two-sum/
// 1. 两数之和
// 使用map key：value， value：idx,
// 首先构建map，然后计算每个值和9的差如果在map中存在就取得idx
func twoSum(nums []int, target int) []int {
	hashNum := map[int]int{}
	for idx, val := range nums {
		diff := target - val
		// 查看是否存在这个差值
		if p, ok := hashNum[diff]; ok {
			return []int{p, idx}
		}
		// 存入当前的val -> idx map pair
		hashNum[val] = idx
	}
	return []int{}
}

// 双指针方法, 先排序，再进行双指针扫描
func twoSum2(nums []int, target int) []int {
	pairs := make([][2]int, 0, len(nums))
	for idx, val := range nums {
		pairs = append(pairs, [2]int{val, idx})
	}
	sort.Slice(pairs, func(i, j int) bool {
		return pairs[i][0] < pairs[j][0]
	})
	r := len(pairs) - 1

	for l := 0; l < len(pairs); l++ {
		for l < r && pairs[l][0]+pairs[r][0] > target {
			r--
		}
		if l < r && pairs[l][0]+pairs[r][0] == target {
			return []int{pairs[l][1], pairs[r][1]}
		}
	}
	return []int{}
}

File: le1.py
"""
python 仅用于练习这个语言，没经过测试
"""
class Solution:
    def twoSum(self, nums, target):
        """
        hashmap 记录preval 对应的index，查看是不是有一个preval和当前值的和是target，
        如果存在就返回preval index 和当前index组成的数组
        """
        hashNum = dict()
        for i, v in enumerate(nums):
            need = target - v
            if need in hashNum: # 是否存在一个preval满足要求
                return [hashNum[need], i] # 返回这个preval的index及当前index组成的数组
            hashNum[v] = i
        return []
    def twoSumDoublePointer(self, nums, target):
        """
        双指针法
        """
        pairs = []
        for i, v in enumerate(nums):
            pairs.append((v,i))
        pairs.sort(key=lambda e:e[0])
        n = len(nums)
        second = n-1
        for first in range(n): # for in 避免 first++
            while first<second  and pairs[first][0] + pairs[second][0] < target:
                second-=1
            if first<second and pairs[first][0] + pairs[second][0] == target:
                return [pairs[first][1], pairs[second][1]]
        return []
